### PR TITLE
Add documentation for PowerFlex Multi Availability Zone

### PR DESCRIPTION
**Pull request by tdawe**
# Description
Add documentation for PowerFlex Multi Availability Zone

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| https://github.com/dell/csm/issues/1612 |

# Checklist:

- [x] Have you run a grammar and spell checks against your submission?
- [x] Have you tested the changes locally?
- [x] Have you tested whether the hyperlinks are working properly?
- [x] Did you add the examples wherever applicable?
- [ ] Have you added high-resolution images?

### CODE REVIEW

File: content/docs/csidriver/features/powerflex.md [hunk 1]
```diff
@@ -1003,3 +1003,84 @@ If such a node is not available, the pods stay in Pending state. This means pods
 Without storage capacity tracking, pods get scheduled on a node satisfying the topology constraints. If the required capacity is not available, volume attachment to the pods fails, and pods remain in ContainerCreating state. Storage capacity tracking eliminates unnecessary scheduling of pods when there is insufficient capacity.
 
 The attribute `storageCapacity.enabled` in `values.yaml` can be used to enable/disable the feature during driver installation using helm. This is by default set to true. To configure how often the driver checks for changed capacity set `storageCapacity.pollInterval` attribute. In case of driver installed via operator, this interval can be configured in the sample file provided [here](https://github.com/dell/csm-operator/blob/main/samples/) by editing the `--capacity-poll-interval` argument present in the provisioner sidecar.
+
+## Multi Availability Zones
```
Comment (AronAtDell) [inline content/docs/csidriver/features/powerflex.md:1007]: Change this to "Multiple Availability Zones"

File: content/docs/csidriver/features/powerflex.md [hunk 2]
```diff
@@ -1003,3 +1003,84 @@ If such a node is not available, the pods stay in Pending state. This means pods
...
+CSI-PowerFlex driver version 2.13.0 and above supports Multi Availability Zones for Block. NFS is not supported at this time.
```
Comment (AronAtDell) [inline content/docs/csidriver/features/powerflex.md:1008]: change to "multiple Availability Zones"

Comment (AronAtDell) [inline content/docs/csidriver/features/powerflex.md:1008]: Change "CSI-PowerFlex driver" to "PowerFlex CSI driver" as you reference it below.

File: content/docs/csidriver/features/powerflex.md [hunk 3]
```diff
@@ -1003,3 +1003,84 @@ If such a node is not available, the pods stay in Pending state. This means pods
...
+
+This feature supports the use a StorageClass that is not associated with any specific PowerFlex system or storage pool. Each cluster node must be labeled with a zone and each PowerFlex system must be assigned to a single zone. When a Pod is scheduled on a node, the volume will be provisioned on the PowerFlex system associated with the cluster node's zone.
+
+Requirements:
+- Every cluster worker node must be labeled with a zone label.
+- Every PowerFlex system in the driver Secret must be assigned to their own zone.
+- The StorageClass does not contain any reference to the storage system id or storage pool.
+- Use the CSM Operator to install the PowerFlex CSI driver. The CSM Operator will detect if Multi Availability Zones are enabled in the driver Secret and ensure the correct MDMs are configured on each worker node during the SDC installation. Currently, helm chart deployment does not support Multi Availability Zones.
```
Comment (AronAtDell) [inline content/docs/csidriver/features/powerflex.md:1016]: "Multi" to "multiple" apply this everywhere in the document.

File: content/docs/csidriver/features/powerflex.md [hunk 4]
```diff
@@ -1003,3 +1003,84 @@ If such a node is not available, the pods stay in Pending state. This means pods
...
+# Label each worker node in the cluster
+kubectl label nodes worker-1 topology.kubernetes.io/zone=zone1
+kubectl label nodes worker-2 topology.kubernetes.io/zone=zone1
+...
+kubectl label nodes worker-3 topology.kubernetes.io/zone=zone2
+kubectl label nodes worker-4 topology.kubernetes.io/zone=zone2
+```
+
+#### StorageClass
+For Multi Availability Zones support, the StorageClass does not require details about the PowerFlex system. Optionally, the `allowedTopologies` can be used to specify topology labels used when provisioning volumes with this StorageClass.
```
Comment (AronAtDell) [inline content/docs/csidriver/features/powerflex.md:1031]: "Multi" to "multiple"

File: content/docs/csidriver/features/powerflex.md (hunk 1 above)
Comment (tdawe) [reply content/docs/csidriver/features/powerflex.md:1007]: Fixed

File: content/docs/csidriver/features/powerflex.md (hunk 2 above)
Comment (tdawe) [reply content/docs/csidriver/features/powerflex.md:1008]: Fixed

Comment (tdawe) [reply content/docs/csidriver/features/powerflex.md:1008]: Fixed

File: content/docs/csidriver/features/powerflex.md (hunk 3 above)
Comment (tdawe) [reply content/docs/csidriver/features/powerflex.md:1016]: Fixed here and everywhere in the document

File: content/docs/csidriver/features/powerflex.md (hunk 4 above)
Comment (tdawe) [reply content/docs/csidriver/features/powerflex.md:1031]: Fixed

File: content/docs/csidriver/features/powerflex.md [hunk 5]
```diff
@@ -1003,3 +1003,84 @@ If such a node is not available, the pods stay in Pending state. This means pods
...
 
 The attribute `storageCapacity.enabled` in `values.yaml` can be used to enable/disable the feature during driver installation using helm. This is by default set to true. To configure how often the driver checks for changed capacity set `storageCapacity.pollInterval` attribute. In case of driver installed via operator, this interval can be configured in the sample file provided [here](https://github.com/dell/csm-operator/blob/main/samples/) by editing the `--capacity-poll-interval` argument present in the provisioner sidecar.
+
+## Multi Availability Zones
+CSI-PowerFlex driver version 2.13.0 and above supports Multi Availability Zones for Block. NFS is not supported at this time.
+
+This feature supports the use a StorageClass that is not associated with any specific PowerFlex system or storage pool. Each cluster node must be labeled with a zone and each PowerFlex system must be assigned to a single zone. When a Pod is scheduled on a node, the volume will be provisioned on the PowerFlex system associated with the cluster node's zone.
```
Comment (atye) [inline content/docs/csidriver/features/powerflex.md:1010]: Should this say "supports the use **of** a StorageClass"?

Comment (tdawe) [reply content/docs/csidriver/features/powerflex.md:1010]: Yes it should, thanks. Fixed

File: content/docs/csidriver/features/powerflex.md [hunk 6]
```diff
@@ -1003,3 +1003,84 @@ If such a node is not available, the pods stay in Pending state. This means pods
...
+
+## Multiple Availability Zones
+PowerFlex CSI driver version 2.13.0 and above supports multiple Availability Zones for Block. NFS is not supported at this time.
+
+This feature supports the use of a StorageClass that is not associated with any specific PowerFlex system or storage pool. Each cluster node must be labeled with a zone and each PowerFlex system must be assigned to a single zone. When a Pod is scheduled on a node, the volume will be provisioned on the PowerFlex system associated with the cluster node's zone.
+
+Requirements:
+- Every cluster worker node must be labeled with a zone label.
+- Every PowerFlex system in the driver Secret must be assigned to their own zone.
+- The StorageClass does not contain any reference to the storage system id or storage pool.
```
Comment (anandrajak1) [inline content/docs/csidriver/features/powerflex.md:1015]: `SystemID` or `storagepool`

File: content/docs/csidriver/features/powerflex.md [hunk 7]
```diff
@@ -1003,3 +1003,84 @@ If such a node is not available, the pods stay in Pending state. This means pods
...
+kubectl label nodes worker-2 topology.kubernetes.io/zone=zone1
+...
+kubectl label nodes worker-3 topology.kubernetes.io/zone=zone2
+kubectl label nodes worker-4 topology.kubernetes.io/zone=zone2
+```
+
+#### StorageClass
+For multiple Availability Zones support, the StorageClass does not require details about the PowerFlex system. Optionally, the `allowedTopologies` can be used to specify topology labels used when provisioning volumes with this StorageClass.
+
+> Note: The StorageClass must use `volumeBindingMode: WaitForFirstConsumer`. Using `volumeBindingMode: Immediate` will not guarantee that the volume is provisioned in the same zone as the scheduled Pod.
```
Comment (anandrajak1) [inline content/docs/csidriver/features/powerflex.md:1033]: **will not guarantee**

File: content/docs/csidriver/features/powerflex.md [hunk 8]
```diff
@@ -1003,3 +1003,84 @@ If such a node is not available, the pods stay in Pending state. This means pods
...
+PowerFlex CSI driver version 2.13.0 and above supports multiple Availability Zones for Block. NFS is not supported at this time.
+
+This feature supports the use of a StorageClass that is not associated with any specific PowerFlex system or storage pool. Each cluster node must be labeled with a zone and each PowerFlex system must be assigned to a single zone. When a Pod is scheduled on a node, the volume will be provisioned on the PowerFlex system associated with the cluster node's zone.
+
+Requirements:
+- Every cluster worker node must be labeled with a zone label.
+- Every PowerFlex system in the driver Secret must be assigned to their own zone.
+- The StorageClass does not contain any reference to the storage system id or storage pool.
+- Use the CSM Operator to install the PowerFlex CSI driver. The CSM Operator will detect if multiple Availability Zones are enabled in the driver Secret and ensure the correct MDMs are configured on each worker node during the SDC installation. Currently, helm chart deployment does not support multiple Availability Zones.
+
```
Comment (anandrajak1) [inline content/docs/csidriver/features/powerflex.md:1017]: > Note: Helm deployment currently doesn’t support multiple Availability Zones.

File: content/docs/csidriver/features/powerflex.md [hunk 9]
```diff
@@ -1003,3 +1003,84 @@ If such a node is not available, the pods stay in Pending state. This means pods
...
+The example manifests below illustrate how to configure two PowerFlex systems, with each system assigned to its own zone. Zone labels can have any custom key, but it must be consistent across the StorageClass, Secret, and Node labels.
+
+#### Labeling Worker Nodes
```
Comment (anandrajak1) [inline content/docs/csidriver/features/powerflex.md:1020]: '###'

File: content/docs/csidriver/features/powerflex.md [hunk 10]
```diff
@@ -1003,3 +1003,84 @@ If such a node is not available, the pods stay in Pending state. This means pods
...
+```
+# Label each worker node in the cluster
+kubectl label nodes worker-1 topology.kubernetes.io/zone=zone1
+kubectl label nodes worker-2 topology.kubernetes.io/zone=zone1
+...
+kubectl label nodes worker-3 topology.kubernetes.io/zone=zone2
+kubectl label nodes worker-4 topology.kubernetes.io/zone=zone2
+```
+
+#### StorageClass
```
Comment (anandrajak1) [inline content/docs/csidriver/features/powerflex.md:1030]: '###'

**Review by anandrajak1 — APPROVED** [review]

File: content/docs/csidriver/features/powerflex.md (hunk 6 above)
Comment (tdawe) [reply content/docs/csidriver/features/powerflex.md:1015]: Fixed

File: content/docs/csidriver/features/powerflex.md (hunk 7 above)
Comment (tdawe) [reply content/docs/csidriver/features/powerflex.md:1033]: Fixed by bolding those words

File: content/docs/csidriver/features/powerflex.md (hunk 8 above)
Comment (tdawe) [reply content/docs/csidriver/features/powerflex.md:1017]: Fixed

**Review by AronAtDell — APPROVED** [review]
Thanks for getting this done!

File: content/docs/csidriver/features/powerflex.md (hunk 10 above)
Comment (tdawe) [reply content/docs/csidriver/features/powerflex.md:1030]: Using 3 hash symbols will add StorageClass as an entry in the table of contents for this section. In my opinion it adds additional clutter to the table of contents. For this reason I'm using 4 hash symbols so the formatting remains the same without adding entries to the table of contents.

File: content/docs/csidriver/features/powerflex.md (hunk 9 above)
Comment (tdawe) [reply content/docs/csidriver/features/powerflex.md:1020]: Using 3 hash symbols will add Labeling Worker Nodes as an entry in the table of contents for this section. In my opinion it adds additional clutter to the table of contents. For this reason I'm using 4 hash symbols so the formatting remains the same without adding entries to the table of contents.

**Review by JacobGros — COMMENTED** [review]
This might be too much of an edge case, but do we want to provide a warning about stale labels? Possibly in troubleshooting section. 

If a node has a topology label like:
```
 csi-vxflexos.dellemc.com/1234=csi-vxflexos.dellemc.com
```
left over from a previous install of the driver, and they install the driver w/ zoning, its possible they'll hit an issue when provisioning a volume since the pod might get assigned to the worker node that has the leftover label. 
This isn't a new issue introduced by zoning, but might make sense to add a warning since now the SDC won't connect to any arrays outside of the zone when zoning is enabled; possibly making this issue more likely.

**Comment by tdawe** [comment]
> This might be too much of an edge case, but do we want to provide a warning about stale labels? Possibly in troubleshooting section.
> 
> If a node has a topology label like:
> 
> ```
>  csi-vxflexos.dellemc.com/1234=csi-vxflexos.dellemc.com
> ```
> 
> left over from a previous install of the driver, and they install the driver w/ zoning, its possible they'll hit an issue when provisioning a volume since the pod might get assigned to the worker node that has the leftover label. This isn't a new issue introduced by zoning, but might make sense to add a warning since now the SDC won't connect to any arrays outside of the zone when zoning is enabled; possibly making this issue more likely.

In this case, a customer would need to have the nodes labeled and the Secret would still contain references to the zone for each PowerFlex system. It does seem like a very small edge case. If the customer removed the zone references from the Secret, then we shouldn't deploy the SDC based on stale labels since we know there aren't any storage systems assigned to those zones.

**Pull request merged**